pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,529 @@
+                base_kernel)
+    return result
+
+
+def _convert_int_g_to_base(int_g: IntG, base_kernel: ExpressionKernel) \
+        -> ExpressionT:
+    """Converts an *IntG* with only one source kernel to an expression with *IntG*s
+    having the base kernel *base_kernel*.
+    """
+    target_kernel = int_g.target_kernel.replace_base_kernel(base_kernel)
+    dim = target_kernel.dim
+
+    result = 0
+
+    density, = int_g.densities
+    source_kernel, = int_g.source_kernels
+    deriv_relation = get_deriv_relation_kernel(source_kernel.get_base_kernel(),
+        base_kernel, hashable_kernel_arguments=(
+            hashable_kernel_args(int_g.kernel_arguments)))
+
+    const = deriv_relation[0]
+    # NOTE: we set a dofdesc here to force the evaluation of this integral
+    # on the source instead of the target when using automatic tagging
+    # see :meth:`pytential.symbolic.mappers.LocationTagger._default_dofdesc`
+    dd = pytential.sym.DOFDescriptor(None,
+            discr_stage=pytential.sym.QBX_SOURCE_STAGE1)
+    const *= pytential.sym.integral(dim, dim-1, density, dofdesc=dd)
+
+    if const != 0 and target_kernel != target_kernel.get_base_kernel():
+        # There might be some TargetPointMultipliers hanging around.
+        # FIXME: handle them instead of bailing out
+        return int_g
+
+    if source_kernel != source_kernel.get_base_kernel():
+        # We assume that any source transformation is a derivative
+        # and the constant when applied becomes zero.
+        const = 0
+
+    result += const
+
+    new_kernel_args = filter_kernel_arguments([base_kernel],
+            int_g.kernel_arguments)
+
+    for mi, c in deriv_relation[1]:
+        knl = source_kernel.replace_base_kernel(base_kernel)
+        for d, val in enumerate(mi):
+            for _ in range(val):
+                knl = AxisSourceDerivative(d, knl)
+                c *= -1
+        result += int_g.copy(source_kernels=(knl,), target_kernel=target_kernel,
+                densities=(density * c,), kernel_arguments=new_kernel_args)
+    return result
+
+
+def get_deriv_relation(kernels: List[ExpressionKernel],
+        base_kernel: ExpressionKernel,
+        kernel_arguments: Mapping[Text, Any],
+        tol: float = 1e-10,
+        order: Optional[int] = None) \
+        -> List[Tuple[ExpressionT, List[Tuple[Tuple[int], ExpressionT]]]]:
+    res = []
+    for knl in kernels:
+        res.append(get_deriv_relation_kernel(knl, base_kernel,
+            hashable_kernel_arguments=hashable_kernel_args(kernel_arguments),
+            tol=tol, order=order))
+    return res
+
+
+@memoize_on_first_arg
+def get_deriv_relation_kernel(kernel: ExpressionKernel,
+        base_kernel: ExpressionKernel,
+        hashable_kernel_arguments: Tuple[Tuple[Text, Any]],
+        tol: float = 1e-10,
+        order: Optional[int] = None) \
+        -> Tuple[ExpressionT, List[Tuple[Tuple[int], ExpressionT]]]:
```

Comment:
```suggestion
        -> Tuple[ExpressionT, List[Tuple[Tuple[int, ...], ExpressionT]]]:
```

Reply:
Thanks. Done in 36d81452ebfdfd726fdf7bec95d5de4710c1f906